File: src/Library/ObjectModel/SoundbridgeCache.vb
''' <summary>
''' Caches List objects so that identical calls to ListXyz return the same
''' object.
''' </summary>
''' <remarks></remarks>
Public Class SoundbridgeCache
   Private Structure CacheKey
      Public Parent As Object
      Public ChildType As Type

      Public Sub New(ByVal parent As SoundbridgeObject, ByVal childType As Type)
         Me.Parent = parent
         Me.ChildType = childType
      End Sub

      Public Overrides Function GetHashCode() As Integer
         Return Parent.GetHashCode() Xor ChildType.GetHashCode()
      End Function

      Public Overrides Function Equals(ByVal obj As Object) As Boolean
         If TypeOf obj Is CacheKey Then
            Dim k As CacheKey = obj
            Return k.ChildType Is ChildType And k.Parent Is Parent
         Else
            Return False
         End If
      End Function
   End Structure

   Private _cache As New Dictionary(Of CacheKey, ISoundbridgeListCacheProvider)()

   Public Function BuildList(Of T As SoundbridgeListObject)(ByVal parent As SoundbridgeObject, ByVal listData As String()) As SoundbridgeObjectCollection(Of T)
      Dim key As New CacheKey(parent, GetType(T))
      Return _cache(key).BuildList(listData)
   End Function

   Public Sub RegisterCache(Of T As SoundbridgeListObject)(ByVal cache As SoundbridgeListCacheProvider(Of T))
      _cache.Add(New CacheKey(cache.Parent, GetType(T)), cache)
   End Sub

   Public Sub DeregisterCache(ByVal parent As SoundbridgeObject, ByVal childType As Type)
      _cache.Remove(New CacheKey(parent, childType))
   End Sub
End Class

Public Interface ISoundbridgeListCacheProvider

   ReadOnly Property ChildType() As Type
   ReadOnly Property Parent() As SoundbridgeObject

   Function BuildList(ByVal listData As String()) As IList

End Interface

Public MustInherit Class SoundbridgeListCacheProvider(Of T As SoundbridgeListObject)
   Implements Pixa.Soundbridge.Library.ISoundbridgeListCacheProvider

   Private _soundbridge As Soundbridge
   Private _parent As SoundbridgeObject
   Private _cache As New SortedList(Of String, T)

   Public Sub New(ByVal soundbridge As Soundbridge, ByVal parent As SoundbridgeObject)
      _parent = parent
      _soundbridge = soundbridge
   End Sub

   Public ReadOnly Property ChildType() As Type Implements ISoundbridgeListCacheProvider.ChildType
      Get
         Return GetType(T)
      End Get
   End Property

   Public ReadOnly Property Parent() As SoundbridgeObject Implements ISoundbridgeListCacheProvider.Parent
      Get
         Return _parent
      End Get
   End Property

   Public ReadOnly Property Soundbridge() As Soundbridge
      Get
         Return _soundbridge
      End Get
   End Property

   Public Function BuildList(ByVal listData As String()) As SoundbridgeObjectCollection(Of T)
      'Remove from the cache any elements not in listData
      Dim sortedListData As String() = listData.Clone()
      Array.Sort(sortedListData)

      Dim listI As Integer = 0
      Dim cacheI As Integer = 0

      While listI < sortedListData.Length And cacheI < _cache.Count
         Dim cache As String = _cache.Keys(cacheI)
         Dim comparison As Integer = String.Compare(_cache.Keys(cacheI), sortedListData(listI))

         If comparison >= 0 Then
            'Either the strings match, or there's a new element in listData
            'So we need to move on to the next list item
            listI += 1
         End If

         If comparison < 0 Then
            'There's an element in the cache that's not in listdata, it needs
            'to be removed, but we don't always want to remove cached objects
            'i.e. A connected media server may not appear in ListServers if it's
            'filtered out.
            'If we don't remove the item, move onto the next element in the cache
            If Not Remove(cacheI) Then
               cacheI += 1
            End If
         ElseIf comparison = 0 Then
            'The strings match, so move on to the next element in the cache.
            cacheI += 1
         End If
      End While

      'Clear out the rest of the items in the cache, bearing in mind that we
      'don't always want to remove an element.
      While cacheI < _cache.Count
         If Not Remove(cacheI) Then
            cacheI += 1
         End If
      End While

      'Build the list, drawing existing elements from listData and creating new elements
      Dim l As SoundbridgeObjectCollection(Of T) = CreateCollection()

      For i As Integer = 0 To listData.Length - 1
         Dim s As String = listData(i)

         'Use the cache if we've seen the element before, otherwise create a new
         'object.  We may need to update the item's index as this is a new list.
         Dim o As T

         If _cache.ContainsKey(s) Then
            o = _cache(s)
            o.Index = i
            l.Add(o)
         Else
            o = CreateObject(s, i)

            If o IsNot Nothing Then
               _cache.Add(s, o)
               l.Add(o)
            End If
         End If
      Next

      Return l
   End Function

   Private Function Remove(ByVal index As Integer) As Boolean
      Dim o As T = _cache.Values(index)

      If o.ShouldCacheDispose Then
         o.Dispose()
         _cache.RemoveAt(index)
         Return True
      Else
         Return False
      End If
   End Function

   Private Function IBuildList(ByVal listData As String()) As IList Implements ISoundbridgeListCacheProvider.BuildList
      Return BuildList(listData)
   End Function

   Protected Overridable Function CreateCollection() As SoundbridgeObjectCollection(Of T)
      Return New SoundbridgeObjectCollection(Of T)(_soundbridge)
   End Function

   Protected MustOverride Function CreateObject(ByVal elementData As String, ByVal index As Integer) As T

End Class

Friend Class SoundbridgeMediaServerCacheProvider
   Inherits Pixa.Soundbridge.Library.SoundbridgeListCacheProvider(Of MediaServer)

   Public Sub New(ByVal soundbridge As Soundbridge, ByVal parent As SoundbridgeObject)
      MyBase.New(soundbridge, parent)
   End Sub

   Protected Overrides Function CreateObject(ByVal elementData As String, ByVal index As Integer) As MediaServer
      Dim tokens As String() = elementData.Split(" ", 3I, StringSplitOptions.None)
      Return New MediaServer(Soundbridge, ServerListAvailabilityToMediaServerAvailability(tokens(0)), ServerListTypeToMediaServerType(tokens(1)), tokens(2), index)
   End Function

   ''' <summary>
   ''' Converts the specified string value into a <see cref="MediaServerAvailability"/>
   ''' value.
   ''' </summary>
   Private Function ServerListAvailabilityToMediaServerAvailability(ByVal value As String) As MediaServerAvailability
      Select Case value
         Case "kOnline"
            Return MediaServerAvailability.Online

         Case "kOffline"
            Return MediaServerAvailability.Offline

         Case "kHidden"
            Return MediaServerAvailability.Hidden

         Case "kInaccessible"
            Return MediaServerAvailability.Inaccessible

      End Select
   End Function

   ''' <summary>
   ''' Converts the specified value into a <see cref="MediaServerType"/>.
   ''' </summary>
   Private Function ServerListTypeToMediaServerType(ByVal value As String) As MediaServerType
      Select Case value
         Case "kITunes"
            Return MediaServerType.Daap

         Case "kUPnP"
            Return MediaServerType.Upnp

         Case "kSlim"
            Return MediaServerType.Slim

         Case "kFlash"
            Return MediaServerType.Flash

         Case "kFavoriteRadio"
            Return MediaServerType.Radio

         Case "kAMTuner"
            Return MediaServerType.AM

         Case "kFMTuner"
            Return MediaServerType.FM

         Case "kRSP"
            Return MediaServerType.Rsp

         Case "kLinein"
            Return MediaServerType.LineIn

         Case Else
            Return -1
      End Select
   End Function
End Class

Friend Class MediaContainerMediaContainerCacheProvider
   Inherits Pixa.Soundbridge.Library.SoundbridgeListCacheProvider(Of MediaContainer)

   Private _listIsContainers As Boolean

   Public Sub New(ByVal soundbridge As Soundbridge, ByVal parent As MediaContainer)
      MyBase.New(soundbridge, parent)
   End Sub

   Protected Overrides Function CreateObject(ByVal elementData As String, ByVal index As Integer) As MediaContainer
      If index = 0 Then
         Dim l As String() = Soundbridge.Client.GetSongInfo(index)
         _listIsContainers = Array.IndexOf(l, "format: unsupported") >= 0
      End If

      If _listIsContainers Then
         Return New MediaContainer(DirectCast(Parent, MediaContainer), index, elementData)
      Else
         Return Nothing
      End If
   End Function
End Class

Friend Class MediaContainerSongCacheProvider
   Inherits Pixa.Soundbridge.Library.SoundbridgeListCacheProvider(Of Song)

   Private _listIsSongs As Boolean

   Public Sub New(ByVal soundbridge As Soundbridge, ByVal parent As MediaContainer)
      MyBase.New(soundbridge, parent)
   End Sub

   Protected Overrides Function CreateCollection() As SoundbridgeObjectCollection(Of Song)
      Return New SongCollection(Soundbridge)
   End Function

   Protected Overrides Function CreateObject(ByVal elementData As String, ByVal index As Integer) As Song
      If index = 0 Then
         Dim l As String() = Soundbridge.Client.GetSongInfo(index)
         _listIsSongs = Array.IndexOf(l, "format: unsupported") < 0
      End If

      If _listIsSongs Then
         Return New Song(DirectCast(Parent, MediaContainer), index, elementData)
      Else
         Return Nothing
      End If
   End Function
End Class

File: src/Pixa.Soundbridge/ObjectModel/SoundbridgeCache.cs
using System;
using System.Collections;
using System.Collections.Generic;

namespace Pixa.Soundbridge.Library
{
    /// <summary>
/// Caches List objects so that identical calls to ListXyz return the same
/// object.
/// </summary>
/// <remarks></remarks>
    public class SoundbridgeCache
    {
        private struct CacheKey
        {
            public object Parent;
            public Type ChildType;

            public CacheKey(SoundbridgeObject parent, Type childType)
            {
                Parent = parent;
                ChildType = childType;
            }

            public override int GetHashCode()
            {
                return Parent.GetHashCode() ^ ChildType.GetHashCode();
            }

            public override bool Equals(object obj)
            {
                if (obj is CacheKey)
                {
                    CacheKey k = (CacheKey)obj;
                    return ReferenceEquals(k.ChildType, ChildType) & ReferenceEquals(k.Parent, Parent);
                }
                else
                {
                    return false;
                }
            }
        }

        private Dictionary<CacheKey, ISoundbridgeListCacheProvider> _cache = new Dictionary<CacheKey, ISoundbridgeListCacheProvider>();

        public SoundbridgeObjectCollection<T> BuildList<T>(SoundbridgeObject parent, string[] listData) where T : SoundbridgeListObject
        {
            var key = new CacheKey(parent, typeof(T));
            return (SoundbridgeObjectCollection<T>)_cache[key].BuildList(listData);
        }

        public void RegisterCache<T>(SoundbridgeListCacheProvider<T> cache) where T : SoundbridgeListObject
        {
            _cache.Add(new CacheKey(cache.Parent, typeof(T)), cache);
        }

        public void DeregisterCache(SoundbridgeObject parent, Type childType)
        {
            _cache.Remove(new CacheKey(parent, childType));
        }
    }

    public interface ISoundbridgeListCacheProvider
    {
        Type ChildType { get; }
        SoundbridgeObject Parent { get; }

        IList BuildList(string[] listData);
    }

    public abstract class SoundbridgeListCacheProvider<T> : ISoundbridgeListCacheProvider where T : SoundbridgeListObject
    {
        private Soundbridge _soundbridge;
        private SoundbridgeObject _parent;
        private SortedList<string, T> _cache = new SortedList<string, T>();

        public SoundbridgeListCacheProvider(Soundbridge soundbridge, SoundbridgeObject parent)
        {
            _parent = parent;
            _soundbridge = soundbridge;
        }

        public Type ChildType
        {
            get
            {
                return typeof(T);
            }
        }

        public SoundbridgeObject Parent
        {
            get
            {
                return _parent;
            }
        }

        public Soundbridge Soundbridge
        {
            get
            {
                return _soundbridge;
            }
        }

        public SoundbridgeObjectCollection<T> BuildList(string[] listData)
        {
            // Remove from the cache any elements not in listData
            string[] sortedListData = (string[])listData.Clone();
            Array.Sort(sortedListData);
            int listI = 0;
            int cacheI = 0;
            while (listI < sortedListData.Length & cacheI < _cache.Count)
            {
                string cache = _cache.Keys[cacheI];
                int comparison = string.Compare(_cache.Keys[cacheI], sortedListData[listI]);
                if (comparison >= 0)
                {
                    // Either the strings match, or there's a new element in listData
                    // So we need to move on to the next list item
                    listI += 1;
                }

                if (comparison < 0)
                {
                    // There's an element in the cache that's not in listdata, it needs
                    // to be removed, but we don't always want to remove cached objects
                    // i.e. A connected media server may not appear in ListServers if it's
                    // filtered out.
                    // If we don't remove the item, move onto the next element in the cache
                    if (!Remove(cacheI))
                    {
                        cacheI += 1;
                    }
                }
                else if (comparison == 0)
                {
                    // The strings match, so move on to the next element in the cache.
                    cacheI += 1;
                }
            }

            // Clear out the rest of the items in the cache, bearing in mind that we
            // don't always want to remove an element.
            while (cacheI < _cache.Count)
            {
                if (!Remove(cacheI))
                {
                    cacheI += 1;
                }
            }

            // Build the list, drawing existing elements from listData and creating new elements
            var l = CreateCollection();
            for (int i = 0, loopTo = listData.Length - 1; i <= loopTo; i++)
            {
                string s = listData[i];

                // Use the cache if we've seen the element before, otherwise create a new
                // object.  We may need to update the item's index as this is a new list.
                T o;
                if (_cache.ContainsKey(s))
                {
                    o = _cache[s];
                    o.Index = i;
                    l.Add(o);
                }
                else
                {
                    o = CreateObject(s, i);
                    if (o is object)
                    {
                        _cache.Add(s, o);
                        l.Add(o);
                    }
                }
            }

            return l;
        }

        private bool Remove(int index)
        {
            var o = _cache.Values[index];
            if (o.ShouldCacheDispose)
            {
                o.Dispose();
                _cache.RemoveAt(index);
                return true;
            }
            else
            {
                return false;
            }
        }

        private IList IBuildList(string[] listData)
        {
            return BuildList(listData);
        }

        IList ISoundbridgeListCacheProvider.BuildList(string[] listData) => IBuildList(listData);

        protected virtual SoundbridgeObjectCollection<T> CreateCollection()
        {
            return new SoundbridgeObjectCollection<T>(_soundbridge);
        }

        protected abstract T CreateObject(string elementData, int index);
    }

    internal class SoundbridgeMediaServerCacheProvider : SoundbridgeListCacheProvider<MediaServer>
    {
        public SoundbridgeMediaServerCacheProvider(Soundbridge soundbridge, SoundbridgeObject parent) : base(soundbridge, parent)
        {
        }

        protected override MediaServer CreateObject(string elementData, int index)
        {
            var tokens = elementData.Split(" ".ToCharArray(), 3, StringSplitOptions.None);
            return new MediaServer(Soundbridge, ServerListAvailabilityToMediaServerAvailability(tokens[0]), ServerListTypeToMediaServerType(tokens[1]), tokens[2], index);
        }

        /// <summary>
   /// Converts the specified string value into a <see cref="MediaServerAvailability"/>
   /// value.
   /// </summary>
        private MediaServerAvailability ServerListAvailabilityToMediaServerAvailability(string value)
        {
            switch (value ?? "")
            {
                case "kOnline":
                    {
                        return MediaServerAvailability.Online;
                    }

                case "kOffline":
                    {
                        return MediaServerAvailability.Offline;
                    }

                case "kHidden":
                    {
                        return MediaServerAvailability.Hidden;
                    }

                case "kInaccessible":
                    {
                        return MediaServerAvailability.Inaccessible;
                    }
            }

            return default;
        }

        /// <summary>
   /// Converts the specified value into a <see cref="MediaServerType"/>.
   /// </summary>
        private MediaServerType ServerListTypeToMediaServerType(string value)
        {
            switch (value ?? "")
            {
                case "kITunes":
                    {
                        return MediaServerType.Daap;
                    }

                case "kUPnP":
                    {
                        return MediaServerType.Upnp;
                    }

                case "kSlim":
                    {
                        return MediaServerType.Slim;
                    }

                case "kFlash":
                    {
                        return MediaServerType.Flash;
                    }

                case "kFavoriteRadio":
                    {
                        return MediaServerType.Radio;
                    }

                case "kAMTuner":
                    {
                        return MediaServerType.AM;
                    }

                case "kFMTuner":
                    {
                        return MediaServerType.FM;
                    }

                case "kRSP":
                    {
                        return MediaServerType.Rsp;
                    }

                case "kLinein":
                    {
                        return MediaServerType.LineIn;
                    }

                default:
                    {
                        return (MediaServerType)(-1);
                    }
            }
        }
    }

    internal class MediaContainerMediaContainerCacheProvider : SoundbridgeListCacheProvider<MediaContainer>
    {
        private bool _listIsContainers;

        public MediaContainerMediaContainerCacheProvider(Soundbridge soundbridge, MediaContainer parent) : base(soundbridge, parent)
        {
        }

        protected override MediaContainer CreateObject(string elementData, int index)
        {
            if (index == 0)
            {
                var l = Soundbridge.Client.GetSongInfo(index);
                _listIsContainers = Array.IndexOf(l, "format: unsupported") >= 0;
            }

            if (_listIsContainers)
            {
                return new MediaContainer((MediaContainer)Parent, index, elementData);
            }
            else
            {
                return null;
            }
        }
    }

    internal class MediaContainerSongCacheProvider : SoundbridgeListCacheProvider<Song>
    {
        private bool _listIsSongs;

        public MediaContainerSongCacheProvider(Soundbridge soundbridge, MediaContainer parent) : base(soundbridge, parent)
        {
        }

        protected override SoundbridgeObjectCollection<Song> CreateCollection()
        {
            return new SongCollection(Soundbridge);
        }

        protected override Song CreateObject(string elementData, int index)
        {
            if (index == 0)
            {
                var l = Soundbridge.Client.GetSongInfo(index);
                _listIsSongs = Array.IndexOf(l, "format: unsupported") < 0;
            }

            if (_listIsSongs)
            {
                return new Song((MediaContainer)Parent, index, elementData);
            }
            else
            {
                return null;
            }
        }
    }
}